{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import argparse\n",
    "import msprime\n",
    "import tskit\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import logging\n",
    "from sklearn.decomposition import PCA\n",
    "from logging import error, warning, info, debug, critical\n",
    "import random\n",
    "from IPython.display import SVG\n",
    "from PyPDF2 import PdfFileMerger\n",
    "\n",
    "\n",
    "import matplotlib.font_manager\n",
    "from IPython.core.display import HTML\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nA = 20  \n",
    "popA = 1000\n",
    "nB = 20 \n",
    "popB = 1000\n",
    "nC = 100\n",
    "popC = 1000\n",
    "\n",
    "nsamps = nA + nB + nC# + nD\n",
    "totpop = popA + popB + popC# + popD\n",
    "\n",
    "\n",
    "p = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)\n",
    "p.add_argument('-v', '--verbose', action='store_true', default = False)\n",
    "p.add_argument('--debug', action='store_true', default = False)\n",
    "p.add_argument('--demdebug', action='store_true', default = False)\n",
    "p.add_argument('--hc', action='store_true', default = False)\n",
    "#Go into msprime.simulate\n",
    "p.add_argument('--seqlen', type=float, default=1e7, help = 'sequence length') \n",
    "p.add_argument('--mu', type=float, default=1e-8, help = 'mutation rate') # default 1e-8\n",
    "p.add_argument('--rho', type=float, default=2e-8, help = 'recombination rate') # default 2e-8\n",
    "\n",
    "\n",
    "p.add_argument('--nsnps', type=int, default=2500, help = 'number of SNPs to use')\n",
    "p.add_argument('--nchrs', type=int, default=1, help = 'number of SNPs to use')\n",
    "\n",
    "#Pruning for PCA\n",
    "p.add_argument('--prune_by_sep', type=float, default=0.0, help = 'min separation between SNPs (otherwise use LD)')\n",
    "p.add_argument('--LD_threshold', type=float, default=0.2, help = 'max LD between SNPs')\n",
    "p.add_argument('--seed', type=int, default=0, help = 'random seed for simulation')\n",
    "#PCA stuff actual\n",
    "p.add_argument('--pcasamples', default=\"{},{},{}\".format(nA,nB,nC), help = 'number of samples to include in PCA from each population: n1,n2,...')\n",
    "p.add_argument('--project_all', action='store_true', default = False, help = 'project all samples onto the PC axes')\n",
    "p.add_argument('--figtitle', default='', help = 'figure title')\n",
    "p.add_argument('--figfmt', default='svg', choices=['svg', 'pdf', 'png'], help = 'figure format')\n",
    "p.add_argument('--figsize', default = '12.0,7.0', help = 'figure size in inches')\n",
    "args = p.parse_known_args()[0]\n",
    "print(args)\n",
    "#this stuff is about checking the code info confirms working as expected, debug gives detailed info for when diagnosing probs\n",
    "loglevel = logging.WARNING\n",
    "if args.verbose:\n",
    "\tloglevel = logging.INFO\n",
    "if args.debug:\n",
    "\tloglevel = logging.DEBUG\n",
    "logging.basicConfig(format = '%(module)s:%(lineno)d:%(levelname)s: %(message)s', level = loglevel)\n",
    "\n",
    "#just generating figure so size input from args.figsize above\n",
    "fsize = [float(x) for x in args.figsize.split(',')]\n",
    "\n",
    "pop_config=[\n",
    "        msprime.PopulationConfiguration(sample_size=nA,initial_size = popA), \n",
    "        msprime.PopulationConfiguration(sample_size=nB,initial_size = popB), \n",
    "        msprime.PopulationConfiguration(sample_size=nC,initial_size = popC)]\n",
    "        \n",
    "if args.pcasamples: \n",
    "\tpcasamples = [int(x) for x in args.pcasamples.split(',')]\n",
    "demevents=[\n",
    "          msprime.MassMigration(500, source = 2, dest = 1, proportion =1),\n",
    "          msprime.MassMigration(1000, source = 1, dest = 0, proportion = 1)]\n",
    "          \n",
    "          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed = round(1000 * random.random())\n",
    "ts = msprime.simulate(population_configurations=pop_config, demographic_events=demevents, length=args.seqlen, mutation_rate=args.mu, recombination_rate=args.rho, random_seed=random_seed)\n",
    "nred = 0\n",
    "a_x = []\n",
    "a_y = []\n",
    "b_x = []\n",
    "b_y = []\n",
    "c_x = []\n",
    "c_y = []\n",
    "\n",
    "\n",
    "for a in range(0,10):\n",
    "    nred = 10*a\n",
    "    \n",
    "\n",
    "    tree = ts.first()\n",
    "    #colour_map = {0:\"red\", 1:\"blue\", 2: \"green\"}\n",
    "    #node_colours = {\n",
    "     #   u: colour_map[tree.population(u)] for u in tree.nodes()} \n",
    "\n",
    "    dd = msprime.DemographyDebugger(\n",
    "        population_configurations=pop_config,\n",
    "        demographic_events=demevents)\n",
    "    if a == 1:\n",
    "        dd.print_history()\n",
    "\n",
    "    variants = np.empty((args.nsnps,nsamps-nred), dtype=\"u1\")\n",
    "    j = 0\n",
    "    if args.prune_by_sep > 0: # currently using LD pruning so this section skipped\n",
    "        info(f'pruning sites by separation > {args.prune_by_sep}')\n",
    "        lastpos = 0\n",
    "        for var in ts.variants():\n",
    "            if lastpos + args.prune_by_sep > args.seqlen:\n",
    "                error(f'ran out of sequence after {j} SNPs')\n",
    "                break\n",
    "            if var.site.position - lastpos > args.prune_by_sep:\n",
    "                debug(j, var.site.id, var.site.position, var.genotypes, sep=\"\\t\")\n",
    "                variants[j] = var.genotypes\n",
    "                j += 1\n",
    "                lastpos = var.site.position\n",
    "            if j == args.nsnps:\n",
    "                break\n",
    "    else:\n",
    "        info(f'LD pruning with threshold {args.LD_threshold}')\n",
    "        ld = tskit.LdCalculator(ts)\n",
    "        next_mutn = 0\n",
    "        tvars = ts.variants()\n",
    "        var = next(tvars)\n",
    "        while j < args.nsnps:\n",
    "            r2 = (ld.r2_array(next_mutn, max_mutations=50) < args.LD_threshold)\n",
    "            if len(r2) == 0 or not np.any(r2):\n",
    "                error(f'ran out of sites after {j} mutations below LD theshold')\n",
    "                break\n",
    "            next_mutn += (1 + np.argmax(r2))\n",
    "            while var.site.id < next_mutn:\n",
    "                var = next(tvars)\n",
    "            var.genotypes = var.genotypes[0:nA+nB+nC -nred]\n",
    "            variants[j] = var.genotypes\n",
    "            debug(f'SNP {j}: {var.site.id} at {var.site.position}')\n",
    "            j += 1\n",
    "        info(f'last SNP {j}: {var.site.id} at {var.site.position}')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    #just for naming PCA plot\n",
    "    # outargs = ' '.join(encmd)\n",
    "    # outname = 'pca' + '_'.join(outargs.split()).replace('_-', '-')\n",
    "\n",
    "\n",
    "    info(f'running PCA')\n",
    "    pcadata = np.ndarray.transpose(variants)\n",
    "    debug(f'data array: {pcadata.shape}')\n",
    "    pca = PCA(n_components=2)\n",
    "    pca.fit(pcadata)\n",
    "\n",
    "    if args.project_all:\n",
    "        samps_pca = pca.transform(data)\n",
    "    else:\n",
    "        samps_pca = pca.transform(pcadata)\n",
    "    \n",
    "\n",
    "    \n",
    "        \n",
    "        \n",
    "    sampa_pca = samps_pca[0:nA]\n",
    "    sampb_pca = samps_pca[nA:nA+nB]\n",
    "    sampc_pca = samps_pca[nA+nB:nA+nB+nC-nred]\n",
    "    #sampd_pca = samps_pca[nA+nB+nC:nA+nB+nC+nD]\n",
    "    #sampe_pca = samps_pca[nA+nB+nC+nD:nA+nB+nC+nD+nE]\n",
    "    #colarr = ['#ee8800', '#ee00ff', '#0033ff', '#00ef00', '#ee0000']\n",
    "\n",
    "\n",
    "    #plt.scatter(samps_pca[:, 0], samps_pca[:, 1])\n",
    "\n",
    "\n",
    "    fig = plt.figure()\n",
    "    plt.xlim(-12,12)\n",
    "    plt.ylim(-6,10)\n",
    "    ax = fig.add_subplot(1,1,1)\n",
    "    ax.scatter(sampa_pca[:, 0], sampa_pca[:, 1], c = '#e41a1c', marker=\"+\", s = 15)\n",
    "    ax.scatter(sampb_pca[:, 0], sampb_pca[:, 1], c = '#377eb8', marker=\"x\", s = 15)\n",
    "    ax.scatter(sampc_pca[:, 0], sampc_pca[:, 1], c = '#4daf4a', marker='o', s =15)\n",
    "    #ax.scatter(sampd_pca[:, 0], sampd_pca[:, 1], c = '#ffc425', marker='p', s =15)\n",
    "    #ax.scatter(sampe_pca[:, 0], sampe_pca[:, 1], c = '#f37735', marker='h', s =15)\n",
    "    ax.scatter(np.mean(sampa_pca[:, 0]), np.mean(sampa_pca[:, 1]), c = '#e41a1c', marker = \"P\", s = 100)\n",
    "    ax.scatter(np.mean(sampb_pca[:, 0]), np.mean(sampb_pca[:, 1]), c = '#377eb8', marker = \"X\", s = 100)\n",
    "    ax.scatter(np.mean(sampc_pca[:, 0]), np.mean(sampc_pca[:, 1]), c = '#4daf4a', marker = \"o\", s = 100)\n",
    "    #ax.scatter(np.mean(sampd_pca[:, 0]), np.mean(sampd_pca[:, 1]), c = '#ffc425', marker = \"p\", s = 100)#f830ff\n",
    "    #ax.scatter(np.mean(sampe_pca[:, 0]), np.mean(sampe_pca[:, 1]), c = '#f37735', marker = \"h\", s = 100)\n",
    "    plt.xlabel(f'PC1', fontname = 'Arial')\n",
    "    plt.ylabel(f'PC2', fontname = 'Arial')\n",
    "    #black_patch = mpatches.Patch(color='#f37735', label='E - 25:75')\n",
    "    plt.legend(('A','B','C'), loc = 'upper right')\n",
    "    #plt.legend(handles=[red_patch,blue_patch,green_patch,purple_patch,black_patch])\n",
    "    if args.figtitle:\n",
    "        plt.title(args.figtitle, fontname = 'Arial', fontsize= 20)\n",
    "        fig.set_size_inches(fsize[0], fsize[1])\n",
    "        plt.subplots_adjust(left=0.15, bottom=0.1, right=1 - 0.05, top= 1 - 0.075)\n",
    "    else:\n",
    "        plt.subplots_adjust(left=0.15, bottom=0.1, right=1 - 0.05, top= 1 - 0.05)\n",
    "\n",
    "    if args.hc:\n",
    "        fig.set_size_inches(fsize[0], fsize[1])\n",
    "        fname = '.'.join([outname + '_' + args.figtitle, args.figfmt])\n",
    "        plt.savefig(fname, format=args.figfmt)\n",
    "        info(f'Saved figure to {fname}')\n",
    "    else:\n",
    "        a_x.append(np.mean(sampa_pca[:, 0]))\n",
    "        a_y.append(np.mean(sampa_pca[:, 1]))\n",
    "        b_x.append(np.mean(sampb_pca[:, 0]))\n",
    "        b_y.append(np.mean(sampb_pca[:, 1]))\n",
    "        c_x.append(np.mean(sampc_pca[:, 0]))\n",
    "        c_y.append(np.mean(sampc_pca[:, 1]))\n",
    "        plt.savefig(\"PCAplot\" + str(a)+ \".pdf\")\n",
    "        plt.show()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdfs = []\n",
    "\n",
    "for a in range(0,9):\n",
    "    pdfs.append('PCAplot'+str(a)+'.pdf')\n",
    "#pdfs = ['PCAplot1.pdf', 'PCAplot2.pdf','PCAplot3.pdf','PCAplot4.pdf','PCAplot5.pdf','PCAplot6.pdf','PCAplot7.pdf','PCAplot8.pdf','PCAplot9.pdf','PCAplot10.pdf','PCAplot11.pdf','PCAplot12.pdf','PCAplot13.pdf','PCAplot14.pdf','PCAplot15.pdf']\n",
    "\n",
    "merger = PdfFileMerger()\n",
    "\n",
    "for pdf in pdfs:\n",
    "    merger.append(pdf)\n",
    "\n",
    "merger.write(\"fig_8_3.pdf\")\n",
    "merger.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
